generic Executor and EvmEventLogger

aathan opened the pull request:
## Motivation

See https://github.com/foundry-rs/foundry/pull/5505

## Solution

aathan commented:
Please look for the NOTE: in crates/evm/src/executor/inspector/logs.rs diff

Evalir commented on crates/anvil/src/eth/backend/mem/mod.rs:
```diff
@@ -884,13 +893,13 @@ impl Backend {
                                         Some(token) => {
                                             node_info!(
-                                                "    Error: reverted with custom error: {:?}",
+                                                "    Error: reverted with custom error (1): {:?}",
```
```suggestion
                                                "    Error: reverted with custom error: {:?}",
```

aathan replied:
FYI the point of this is to differentiate the source of the error. anvil's errors are sometimes formatted exactly the same but originate in different places.

Evalir commented on crates/anvil/src/eth/backend/mem/mod.rs:
```diff
@@ -884,13 +893,13 @@ impl Backend {
                                         None => {
                                             node_info!(
-                                                "    Error: reverted with custom error: {}",
+                                                "    Error: reverted with custom error (2): {}",
```
```suggestion
                                                "    Error: reverted with custom error: {}",
```

aathan replied:
FYI the point of this is to differentiate the source of the error. anvil's errors are sometimes formatted exactly the same but originate in different places.

Evalir commented on crates/evm/src/executor/inspector/logs.rs:
```diff
@@ -82,7 +113,20 @@ const TOPIC: H256 = H256([
 fn convert_hh_log_to_event(call: HardhatConsoleCalls) -> Log {
     // Convert the parameters of the call to their string representation using `ConsoleFmt`.
-    let fmt = call.fmt(Default::default());
+    let fmt = foundry_macros::ConsoleFmt::fmt(&call, Default::default());
+    /* NOTE:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+       In the master branch this is call.fmt(Default::default())
+       However, if I change that here, I get the below error.
+       I don't understand why, as the change seems self contained, and if I'm getting an error whereas no such
+       error is generated here, in the master branch ... that implies there's some incompatible diff happeneing
+       elsewhere ... a diff that shouldn't exist! Yikes.
+
+        error[E0277]: the trait bound `&mut std::fmt::Formatter<'_>: std::default::Default` is not satisfied
+       --> crates/evm/src/executor/inspector/logs.rs:117:24
+        |
+    117 |     let fmt = call.fmt(Default::default());
+        |                        ^^^^^^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `&mut std::fmt::Formatter<'_>`
+         */
```
This is fine—it's just a warning about an unimplemented trait—if it works like this it's aight!

Evalir commented on crates/evm/src/executor/mod.rs:
```diff
@@ -88,11 +89,25 @@ pub struct Executor {
 // === impl Executor ===
 
-impl Executor {
+// the Clone derivation ends up producing errors elsewhere (expected Executor but got &Executor)
```
wdym with this?

Evalir reviewed (commented):
I have to review this a bit more carefully, as it's a design change, but overall it doesn't seem unreasonable. I'd love your eyes @mattsse as well!

aathan commented:
@Evalir I incorporated your requests. Listen, I've spent way too much time rebasing this pull request over and over. Please go ahead and make any trivial changes or comment suggestions you have and either accept or reject the PR. Waiting for me to respond to fairly trivial comments and having 40 commits happen in the meantime, such that I have to review all the deltas again is just not worth it.

As I mentioned in the original PR this is really about causing Executor to be generic, and having spent a little more time with the code, it's really `InspectorStack` itself that probably should be the generic type variable. ONLOG after all is basically looking to modify the per-call behavior of one of the inspectors in the InspectorStack.

The new commits I pulled show that you guys have now added an ExecutorBuilder. And really there already was what amounted to an InspectorStackBuilder (that's what InspectorStackConfig.stack() was doing) ...

What I'm getting at is that the specific inspectors and what their behaviors are is an integral part of definition the Executor, and what the specific behavior of each inspector in the inspector stack is, on a per tool basis (anvil, chisel, etc) should probably be embodied in a generic InspectorStack<... inspectors...>, which in my instant PR takes only the ONLOG generic which in turn modifies only the EvmEventLogger (formerly LogCollector).

The reason the current (and former) design is good 'nuf is that all the tools tend to share inspectors that all behave the same way across all the tools.

Evalir commented:
Hey, so, understood. Will close this PR for now—I think these are changes we definitely want, but we're undergoing a types migration and rebasing will just be a pain for either of us. Definitely agree this is a step in the right direction, and what we'll end up doing is probably getting these changes in after the migration.

Thanks for the contrib!
